File: server/filters/funnel.c
#include "filters.h"

FILTER_SETUP( funnel )
{
	callback->args.funnel.channel = atoi( args_name ) - 1 & 0x0f;
	printf("funnel %s", args_name);
}
FILTER_CALLBACK( funnel )
{
	snd_rawmidi_t *port = data->output_device->midi;
	unsigned char target = (unsigned char) args->funnel.channel;
	unsigned int mask = MASK_ALL; // any channel message is fine
	unsigned int current_mask = 0;
	if( data->output_device->midi_in_exclusive == data->midi_in )
		current_mask = MASK_SYSEX;

	int a = 0;
	for( int b = 0; b < n_bytes; ++b )
	{
		unsigned char cur_state = scan_status(data, buf[b]);
		if( buf[b] >= 0xf8 )
			current_mask = MASK_RT;
		else if( buf[b] >= 0xf0 )
		{
			if( buf[b] == 0xf0 )
				data->output_device->midi_in_exclusive = data->midi_in;
			else if( buf[b] == 0xf7 )
				data->output_device->midi_in_exclusive = NULL;
			current_mask = MASK_SYSEX;
		}
		else if( buf[b] >= 0x80 )
		{
			current_mask = 2 << (buf[b] & 0x0f);
			out_buf[a++] = buf[b] & 0xf0 | target;
			continue;
		}

		if( mask & current_mask )
		{
			out_buf[a++] = buf[b];
		}
	}
	return write_buffer( port, out_buf, a );
}

File: rust/src/filters/funnel.rs

pub struct Funnel { channel : u8 }
impl Funnel {
	pub fn new( args : String ) -> Funnel {
		Funnel { channel : parse_hex(&args) }
	}
}
impl CallbackFn for Funnel {
	fn callback(&self, data : &mut CallbackData, buf : &[u8]) -> usize {
		let mask = MASK_ALL;
		let mut current_mask = 0;

		if data.output_device.midi_in_exclusive == data.midi_in {
			current_mask = MASK_SYSEX;
		}

		let mut a = 0;
		for c in buf {
			let _cur_state = data.output_device.scan_status(*c);
			if *c >= 0xf8 {
				current_mask = MASK_RT;
			} else if *c >= 0xf0 {
				if *c == 0xf0 {
					data.output_device.midi_in_exclusive = data.midi_in.clone();
				} else if *c == 0xf7 {
					data.output_device.midi_in_exclusive = "".to_string();
				}
				current_mask = MASK_SYSEX;
			} else if *c >= 0x80 {
				current_mask = 2 << (*c & 0x0f);
				data.buffer[a] = *c & 0xf0 | self.channel;
				a += 1;
				continue;
			}

			if mask & current_mask != 0 {
				data.buffer[a] = *c;
				a += 1;
			}
		}
		data.output_device.send_buffer(&data.buffer[0..a]).unwrap()
	}
}
